specs/math.txt

```diff
@@ -0,0 +1,127 @@
+Run this with `cargo run -- -M -s specs/math.txt`.
+
+Mathematical expressions extension.
+See the following link to know the syntax.
+https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions
+
+Inline-level mathematical expressions:
+
+```````````````````````````````` example
+This sentence uses `$` delimiters to show math inline:  $\sqrt{3x-1}+(1+x)^2$
+$\sum_{k=1}^n a_k b_k$: Mathematical expression at head of line
+
+`\` may follow just after the first `$`: $\{1, 2, 3\}$
+.
+<p>This sentence uses <code>$</code> delimiters to show math inline:  <span class="math inline">\sqrt{3x-1}+(1+x)^2</span>
+<span class="math inline">\sum_{k=1}^n a_k b_k</span>: Mathematical expression at head of line</p>
+<p><code>\</code> may follow just after the first <code>$</code>:  <span class="math inline">\{1, 2, 3\}</span>
+````````````````````````````````
+
+Block-level mathematical expressions:
+
+```````````````````````````````` example
+**The Cauchy-Schwarz Inequality**
+
+$$\left( \sum_{k=1}^n a_k b_k \right)^2 \leq \left( \sum_{k=1}^n a_k^2 \right) \left( \sum_{k=1}^n b_k^2 \right)$$
+.
+<p><strong>The Cauchy-Schwarz Inequality</strong></p>
+<div class="math block">\left( \sum_{k=1}^n a_k b_k \right)^2 \leq \left( \sum_{k=1}^n a_k^2 \right) \left( \sum_{k=1}^n b_k^2 \right)</div>
+````````````````````````````````
+
+Empty expressions are edge cases that are not parsed as mathematical expressions:
+
+```````````````````````````````` example
+Oops empty $$ expression.
+
+$$$$
+
+This is also empty: $``$
+.
+<p>Oops empty $$ expression.</p>
+<p>$$$$</p>
+<p>This is also empty: $``$</p>
+````````````````````````````````
+
+Sole `$` character is handled as normal text. And Inline-level mathematical expressions cannot continue across newlines:
+
+```````````````````````````````` example
+Hello $ world.
+
+Dollar at end of line: $
+.
+<p>Hello $ world.</p>
+<p>Dollar at end of line: $</p>
+````````````````````````````````
```

Comment:
Great to see more activity around this long-standing issue! I'm willing to work on this if any help is needed.

On this part, I'm not sure I agree. In LaTeX and markdown files alike, a common practice is to justify paragraphs to a width of e.g. 80 characters. Single newlines do not affect the output, even inside math expressions.

There seems to be a divergence of behavior here between markdown files and issues/comments on GitHub. Comments preserve newlines from the raw text, but in markdown files this is not the case. This markup:

```
text and $math broken
into multiple lines$
```

renders in this comment as

text and $math broken
into multiple lines$

but is shown as math in a Gist: https://gist.github.com/ollpu/2b0e18b8fe20d76b97d9e89912f4df3e

It also works with [Pandoc](https://pandoc.org/try/).

Since with pulldown-cmark single newlines are generally ignored, it would make sense to ignore them inside inline math as well.

There is subtlety with how newlines can start/end a math expression though. GitHub and Pandoc have different behavior: On GitHub, an inline math expression cannot start with a newline right after the `$`, but it can end with one before the closing `$`. (See the [same Gist](https://gist.github.com/ollpu/2b0e18b8fe20d76b97d9e89912f4df3e) for examples of this) Pandoc is more consistent, and inline math can neither start or end with a newline – they are essentially treated the same as spaces for this purpose.

Escaped newlines (newline preceded by `\`) are another can of worms.

For what it's worth, I prefer following Pandoc over GitHub in this and other similar cases since the implementation is more mature, consistent and doesn't suffer from the robustness issues that GitHub did at least originally.

Reply:
Similarly, I support following $\LaTeX$'s standards. They are around for a good while and solid.

Reply:
Another common thing supported by both GitHub are display-mode math blocks inside a paragraph.

```
display $$\LaTeX$$ math
```

renders as

display $$\LaTeX$$ math

Reply:
- `$` doesn't continue across lines. In this test case, there is no math expression.
- This PR refers GitHub's markdown renderer behavior. It may differs from LaTeX behavior but it is intended
- I know several edge cases are existing but I intentionally didn't implement it since they make implementation more complex and harder to maintain

Reply:
I'm eyeing a refactor to this PR branch to make math parsing kind of combine the logic of `ItemBody::MaybeEmphasis` and `ItemBody::MaybeCode`. #477 does something similar for inline math, but not display blocks. It seems like display math should behave a little closer to display math and inline code, but just not do the whitespace checks.

What do you think @rhysd?

Reply:
I went ahead and implemented that. More details in the PR: https://github.com/rhysd/pulldown-cmark/pull/1

I'd rather not have a third competing PR to upstream, so hopefully we can converge on a good design here.

Reply:
> I found another bug: When you want to parse this markdown:
> 
> ```
> math
> $$x$$
> ```
> 
> This is not interpreted as math. I tested it with the latest version:
> 
> ```shell
> $ echo -e 'math \n$$x$$' | cargo run -- -m -e
>     Finished dev [unoptimized + debuginfo] target(s) in 0.04s
>      Running `target/debug/pulldown-cmark -m -e`
> 0..13: Start(Paragraph)
> 0..5: Text(Borrowed("math "))
> 5..6: SoftBreak
> 7..9: Text(Borrowed("$$"))
> 9..10: Text(Borrowed("x"))
> 10..12: Text(Borrowed("$$"))
> 0..13: End(Paragraph)
> EOF
> ```

I don't know if anybody saw this, but it still doesn't work.
I'm pretty sure it is not the default behaviour for most math-in-markdown implementations.

Reply:
@rambip 

Yeah, to be clear, I made a separate PR onto @rhysd's branch, but that is not incorporated here yet.

This case works with [my version](https://github.com/ollpu/pulldown-cmark/tree/alt-math) there though:

```bash
$ echo -e 'math \n$$x$$' | cargo run -- -m -e
   Compiling pulldown-cmark v0.9.2 (/home/ollpu/git/pulldown-cmark)
    Finished dev [unoptimized + debuginfo] target(s) in 3.77s
     Running `target/debug/pulldown-cmark -m -e`
0..12: Start(Paragraph)
0..5: Text(Borrowed("math "))
5..6: SoftBreak
6..11: Math(Display, Borrowed("x"))
0..12: End(Paragraph)
EOF
```

Reply:
Ok nice !

Reply:
I'm sorry for the delay. I didn't have enough time to see each PRs I've created this week.

> There seems to be a divergence of behavior here between markdown files and issues/comments on GitHub.

I didn't know this. Thank you for your PR. I'll take a look.